src/ophyd_async/core/signal.py

```diff
@@ -253,6 +257,52 @@ def set_sim_callback(signal: Signal[T], callback: ReadingValueCallback[T]) -> No
     return _sim_backends[signal].set_callback(callback)
 
 
+async def verify_readable(
+    func: Callable[[], dict[str, Any] | Awaitable[dict[str, Any]]],
+    expectation: Dict[str, Any],
+) -> None:
+    """Loop thorough dictionary and verify it against
+    a expected dictionary (readable)"""
+    if asyncio.iscoroutinefunction(func):
+        result = await func()
+    else:
+        result = func()
+
+    for signal in expectation:
+        for field in expectation[signal]:
+            if field == "timestamp":
+                assert isinstance(result[signal]["timestamp"], float)
+            else:
+                assert result[signal][field] == expectation[signal][field]
+
+
+async def assert_value(signal: SignalR[T], value: T) -> None:
+    """Assert a signal value"""
+    assert await signal.get_value() == value
+
+
+async def assert_reading(
+    readable: Readable, reading: Dict[str, Reading] | dict[str, dict[str, Any]]
```

Comment:
I think we should explicitly require readings and not allow the `dict[str, Any]` path, ditto for `assert_configuration`. Also, I recommend using the `Mapping` type rather than `Dict` if you do not expect the dictionary to be mutated

Reply:
I have pretty much made all the changes suggested, just want to make a note of things i noticed making this change, By forcing Mapping t[str,Reading], mypy will fail unless Reading is in the correct order of {value, timestamp, alarm_severity, message} e.g. 
```
 {
                "value": 5.0,
                "timestamp": ANY,
                "alarm_severity": 0,
                "message": "str",
}
```
This also make [varify](https://github.com/DiamondLightSource/dodal/blob/6b43b48fcb388c43f919b7d0866fb44d8f8199ee/tests/devices/unit_tests/test_synchrotron.py#L201) function seem less versatile, It will still behave in the same way but mypy will fail if a function does not return Mapping[str,Reading].

Reply:
I'd put it back to dict then, we don't want to enforce this ordering...

Reply:
This seems like weird behaviour from `TypedDict` and I can't reproduce it, maybe something else is going on here? https://mypy-play.net/?mypy=latest&python=3.12&gist=57a93d787c3cbfe8ab39c061e398edde

Reply:
After merging with main I can no longer create this behaviour, so nothing to worry about.